pkg/utils/radish/scheduler.go

```diff
@@ -0,0 +1,234 @@
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/rs/zerolog"
+)
+
+// Scheduler manages a list of future tasks and on or after the time that they are
+// supposed to be scheduled, the scheduler sends the task on the out channel. This
+// allows radish to schedule tasks for the future or to retry tasks with backoff delays
+// so that the workers are not overwhelmed by long running tasks.
+//
+// The scheduler is implemented with a sorted list of Futures (basically tasks arranged
+// in time order) along with a single go routine that sleeps until the timestamp of the
+// next task unless interrupted by a newly scheduled task. The goal of the scheduler is
+// to minimize the number of go routines and CPU cycles in use so that higher priority
+// work by the task manager or the main go routine is favored by the CPU. To that end
+// the scheduler does not use a "ticker" clock checking if it should execute every
+// second, perferring longer sleeps and interrupts instead.
+type Scheduler struct {
+	sync.RWMutex
+	out     chan<- Task
+	add     chan *Future
+	stop    chan struct{}
+	tasks   Futures
+	running bool
+	logger  zerolog.Logger
+}
+
+// Create a new scheduler that can schedule task futures. The out channel is used to
+// dispatch tasks at their scheduled time. If a task is sent on the out channel it means
+// that the task should be executed as soon as possible. The scheduler makes no
+// guarantees about exact timing of tasks scheduled except that the task will not be
+// sent on the out channel before its scheduled time.
+func NewScheduler(out chan<- Task, logger zerolog.Logger) *Scheduler {
+	return &Scheduler{
+		out:     out,
+		add:     make(chan *Future, 1),
+		stop:    make(chan struct{}),
+		tasks:   make(Futures, 0, minFuturesCapacity),
+		running: false,
+		logger:  logger,
+	}
+}
+
+// Delay schedules the task to be run on or after the specified delay duration from now.
+func (s *Scheduler) Delay(delay time.Duration, task Task) error {
+	return s.Schedule(time.Now().Add(delay), task)
+}
+
+// Schedule a task to run on or after the specified timestamp. If the scheduler is
+// running the task future is sent to the main channel loop, otherwise the tasks is
+// simply inserted into the futures slice. Schedule blocks until the task is received
+// by the main scheduler loop.
+func (s *Scheduler) Schedule(at time.Time, task Task) error {
+	future := &Future{Time: at, Task: task}
+
+	s.Lock()
+	if s.running {
+		s.add <- future
+	} else {
+		s.tasks = s.tasks.Insert(future)
+	}
+	s.Unlock()
+	return nil
+}
+
+// Start the scheduler in its own go routine or no-op if already started.
+func (s *Scheduler) Start() {
+	s.Lock()
+	defer s.Unlock()
+	if s.running {
+		return
+	}
+
+	s.running = true
+	go s.run()
+}
+
+// Run the scheduler, or a no-op if it is already running.
+func (s *Scheduler) Run() {
+	s.Lock()
+	if s.running {
+		s.Unlock()
+		return
+	}
+
+	s.running = true
+	s.Unlock()
+	s.run()
+}
+
+func (s *Scheduler) run() {
```

Comment:
The nice thing about this scheduler is that it doesn't check every second if  tasks need to be run, it uses a timer to wait until the next task needs to be run and can be interrupted if a new task is added before that task.